packages/terra-paginator/src/ProgressivePaginator.jsx

```diff
@@ -0,0 +1,95 @@
+import React from 'react';
+import PropTypes from 'prop-types';
+import classNames from 'classnames/bind';
+import IconNext from 'terra-icon/lib/icon/IconNext';
+import IconPrevious from 'terra-icon/lib/icon/IconPrevious';
+import NumberField from 'terra-form/lib/NumberField';
+
+import 'terra-base/lib/baseStyles';
+import styles from './Paginator.scss';
+
+import { calculatePages } from './_paginationUtils';
+
+const cx = classNames.bind(styles);
+
+const propTypes = {
+  /**
+   * Function to be executed when a navigation element is selected.
+   */
+  onPageChange: PropTypes.func.isRequired,
+  /**
+   * The active/selected page.
+   */
+  selectedPage: PropTypes.number.isRequired,
+  /**
+   * Total number of all items being paginated.
+   */
+  totalCount: PropTypes.number.isRequired,
+  /**
+   * Total number of items per page.
+   */
+  itemCountPerPage: PropTypes.number.isRequired,
+};
+
+class ProgressivePaginator extends React.Component {
+  constructor(props) {
+    super(props);
+
+    const { selectedPage, totalCount } = this.props;
+
+    this.state = {
+      pageInput: selectedPage,
+      selectedPage: selectedPage && totalCount ? selectedPage : undefined,
+    };
+
+    this.handlePageChange = this.handlePageChange.bind(this);
+    this.handlePageInput = this.handlePageInput.bind(this);
+    this.buildPageSelectorForm = this.buildPageSelectorForm.bind(this);
+  }
+
+  handlePageInput(e) {
+    this.setState({ pageInput: e.target.value });
+  }
+
+  handlePageChange(index) {
+    return (event) => {
+      event.preventDefault();
+
+      this.props.onPageChange(event, index);
+      this.setState({ pageInput: index, selectedPage: index });
+    };
+  }
+
+  buildPageSelectorForm(totalPages, onSubmit) {
+    const { pageInput } = this.state;
+
+    const pageForm = (<div>
+      <form className={cx('page-form')} onSubmit={onSubmit(parseInt(pageInput, 10))}>
+        <div>Page {<NumberField isInline value={pageInput} onChange={this.handlePageInput} min={1} max={totalPages} />} of {totalPages}</div>
```

Comment:
The visuals on this are a bit off compared to the callouts:
Current visual:
<img width="148" alt="screen shot 2018-04-30 at 8 43 19 am" src="https://user-images.githubusercontent.com/633148/39430118-8dbde29e-4c52-11e8-97e9-883c5e1ad2e9.png">

Callouts:
<img width="466" alt="screen shot 2018-04-30 at 8 41 45 am" src="https://user-images.githubusercontent.com/633148/39430101-7f5add60-4c52-11e8-8f90-b87456ab2b2e.png">

Setting the wrapping div to `display: flex; align-items: baseline;` should help better vertically align the text with the input. That will need to be tested cross-browser. We may also need to add custom styles to the input to better control the width of it.

Reply:
https://github.com/cerner/terra-core/pull/1445/commits/c619c108cd3f92c3084a93d5dc3d99e53bd717ba

Note: The number appears off-center because of the number-spinner that webkit adds to number inputs.